Add support for NAV-TIMENAVIC message

alinsavix opened the pull request:
## Description

...as documented in UBX-21048598 rev R01. There are probably other NAVIC things that need adding as well, but that is a matter for another day.

This message is provided by, at a minimum, the F9T series of receivers.

## Testing

Test data provided direct from F9T hardware and parsing has been hand-verified as correct. Stream test in `tests/test_stream.py` updated and passes.

## Checklist:

- [x] My code follows the style guidelines of this project (see `CONTRIBUTING.MD`).
- [x] I have performed a self-review of my own code.
- [x] I have cited my u-blox documentation source(s).
- [ ] (N/A) I have commented my code, particularly in hard-to-understand areas.
- [ ] (N/A) I have made corresponding changes to the documentation.
- [x] I have added test cases to the `tests/test_*.py` unittest suite to maintain >= 99% code coverage.
- [x] I have tested my code against the full `tests/test_*.py` unittest suite.
- [x] My changes generate no new warnings.
- [ ] (N/A) Any dependent changes have been merged and published in downstream modules.
- [x] I understand and acknowledge that the code will be published under a BSD 3-Clause license.

semuadmin reviewed (changes requested):
Many thanks for this - much appreciated - but can we add the NAV2-NAVIC message (b"\x29\x63") at the same time? The payload definition is the same as NAV-NAVIC.

The other thing you need to check when adding new message definitions is the configuration database - it's toward the bottom of the Interface Specification. There may be additional CFG-MSGOUT-UBX_NAV databases entries to be added to `ubxtypes_configdb.py` - in this case there are 10 additional entries (_if you cut-and-paste them from the interface specification, remember to change any hyphens into underscores_):

CFG_MSGOUT_UBX_NAV_TIMENAVIC_I2C 0x209106a2 U1
CFG_MSGOUT_UBX_NAV_TIMENAVIC_SPI 0x209106a6 U1
CFG_MSGOUT_UBX_NAV_TIMENAVIC_UART1 0x209106a3 U1
CFG_MSGOUT_UBX_NAV_TIMENAVIC_UART2 0x209106a4 U1
CFG_MSGOUT_UBX_NAV_TIMENAVIC_USB 0x209106a5 U1

CFG_MSGOUT_UBX_NAV2_TIMENAVIC_I2C 0x209106a7 U1
CFG_MSGOUT_UBX_NAV2_TIMENAVIC_SPI 0x209106ab U1
CFG_MSGOUT_UBX_NAV2_TIMENAVIC_UART1 0x209106a8 U1
CFG_MSGOUT_UBX_NAV2_TIMENAVIC_UART2 0x209106a9 U1
CFG_MSGOUT_UBX_NAV2_TIMENAVIC_USB 0x209106aa U1

If you can make these additional changes that would be much appreciated.

alinsavix commented:
Ah, yeah, good call on the NAV2 version of the message. I wonder how many of the definitions are 100% identical between NAV and NAV2, and if there's some sort of easy way to have it do the right thing in those cases, without having to explicitly spell it out every time...

...I was trying to be lazy and not care about the config messages yet, but now that you've gone and called me out on it, I have no choice but to implement them!

I'll get it taken care of in the next day or two.

semuadmin commented:
Hi @alinsavix AFAICS all the NAV2 payloads are identical to the NAV ones and I had considered cross-referencing them rather than duplicating the code (I do this for many of the CFG GET/SET messages), but for the time being I'm working on the assumption that they could diverge at some point.

alinsavix commented:
Alright, I think I got everything in the right places now. I didn't add tests for the NAV2 version, nor the MSGOUT configs, since it doesn't look like you're testing those currently (or, rather, you do a generic test for the MSGOUT configs, but not for a complete set of possible messages, which makes sense given how standard the formatting is on MSGOUT things.

alinsavix commented:
FWIW, it might be useful to add "if I added a NAV message I added the NAV2 counterpart" and "if I added a new message I also updated the MSGOUT configs for that message" to the checklist in the PR template.

semuadmin commented:
Hi @alinsavix Thanks again for this. You may have noticed that the GitHub Actions workflow tasks failed for this merge (`build: failing`), but that was (_another_) temporary bug in GitHub Actions and nothing to do with your changes. It appears to be resolved now.
